From: Danilo Egea Gondolfo <danilo.egea.gondolfo@canonical.com>
Subject: dbus: Build the copy path correctly

This patch fixes a bug in the DBus integration where it was silently failing to
copy the netplan configuration files due to a malformed file path.
It also improves the error handling in netplan-dbus by propagating the failure to
the function callers.

This fix is related to https://github.com/canonical/netplan/pull/299

Origin: upstream, https://github.com/canonical/netplan/commit/2ff1daa9fcdb5779bfb88631a0af1104a0722f1e and https://github.com/canonical/netplan/commit/a76739110da77920f73d467e4102df2eb0d1af2c
Bug-Ubuntu: https://bugs.launchpad.net/netplan/+bug/1997467
---
 src/dbus.c | 21 +++++++++++++--------
 1 file changed, 13 insertions(+), 8 deletions(-)

diff --git a/src/dbus.c b/src/dbus.c
index 7f633e6..a5125fc 100644
--- a/src/dbus.c
+++ b/src/dbus.c
@@ -141,7 +141,7 @@ _copy_yaml_state(char *src_root, char *dst_root, sd_bus_error *ret_error)
     gchar *dest_path = NULL;
     size_t len = strlen(src_root);
     for (size_t i = 0; i < gl.gl_pathc; ++i) {
-        dest_path = g_strjoin(NULL, dst_root, (gl.gl_pathv[i])+len, NULL);
+        dest_path = g_build_path(G_DIR_SEPARATOR_S, dst_root, (gl.gl_pathv[i])+len, NULL);
         source = g_file_new_for_path(gl.gl_pathv[i]);
         dest = g_file_new_for_path(dest_path);
         g_file_copy(source, dest, G_FILE_COPY_OVERWRITE
@@ -223,8 +223,8 @@ _backup_global_state(sd_bus_error *ret_error)
     }
 
     /* Copy main *.yaml files from /{etc,run,lib}/netplan/ to GLOBAL backup dir */
-    _copy_yaml_state(NETPLAN_ROOT, path, ret_error);
-    return 0;
+    r = _copy_yaml_state(NETPLAN_ROOT, path, ret_error);
+    return r;
 }
 
 /**
@@ -444,7 +444,8 @@ netplan_try_cancelled_cb(sd_event_source *es, const siginfo_t *si, void* userdat
         unlink_glob(NETPLAN_ROOT, "/{etc,run,lib}/netplan/*.yaml");
         /* Restore GLOBAL backup config state to main rootdir */
         state_dir = g_strdup_printf("%s/run/netplan/config-%s", NETPLAN_ROOT, NETPLAN_GLOBAL_CONFIG);
-        _copy_yaml_state(state_dir, NETPLAN_ROOT, NULL);
+        r = _copy_yaml_state(state_dir, NETPLAN_ROOT, NULL);
+        if (r < 0) return r;
 
         /* Un-invalidate all other current config objects */
         if (!g_strcmp0(d->handler_id, d->config_dirty))
@@ -564,7 +565,8 @@ method_config_apply(sd_bus_message *m, void *userdata, sd_bus_error *ret_error)
         unlink_glob(NETPLAN_ROOT, "/{etc,run,lib}/netplan/*.yaml");
         /* Copy current config state to GLOBAL */
         state_dir = g_strdup_printf("%s/run/netplan/config-%s", NETPLAN_ROOT, d->config_id);
-        _copy_yaml_state(state_dir, NETPLAN_ROOT, ret_error);
+        r = _copy_yaml_state(state_dir, NETPLAN_ROOT, ret_error);
+        if (r < 0) return r;
         d->handler_id = g_strdup(d->config_id);
     }
 
@@ -639,7 +641,8 @@ method_config_try(sd_bus_message *m, void *userdata, sd_bus_error *ret_error)
 
     /* Copy current config *.yaml state to main rootdir (i.e. /etc/netplan/) */
     state_dir = g_strdup_printf("%s/run/netplan/config-%s", NETPLAN_ROOT, d->config_id);
-    _copy_yaml_state(state_dir, NETPLAN_ROOT, ret_error);
+    r = _copy_yaml_state(state_dir, NETPLAN_ROOT, ret_error);
+    if (r < 0) return r;
 
     /* Exec try */
     r = method_try(m, userdata, ret_error);
@@ -670,7 +673,8 @@ method_config_cancel(sd_bus_message *m, void *userdata, sd_bus_error *ret_error)
         unlink_glob(NETPLAN_ROOT, "/{etc,run,lib}/netplan/*.yaml");
         /* Restore GLOBAL backup config state to main rootdir */
         state_dir = g_strdup_printf("%s/run/netplan/config-%s", NETPLAN_ROOT, NETPLAN_GLOBAL_CONFIG);
-        _copy_yaml_state(state_dir, NETPLAN_ROOT, ret_error);
+        r = _copy_yaml_state(state_dir, NETPLAN_ROOT, ret_error);
+        if (r < 0) return r;
 
         /* Clear GLOBAL backup and config state */
         _clear_tmp_state(NETPLAN_GLOBAL_CONFIG, d);
@@ -754,7 +758,8 @@ method_config(sd_bus_message *m, void *userdata, sd_bus_error *ret_error)
     }
 
     /* Copy all *.yaml files from /{etc,run,lib}/netplan/ to temp dir */
-    _copy_yaml_state(NETPLAN_ROOT, path, ret_error);
+    r = _copy_yaml_state(NETPLAN_ROOT, path, ret_error);
+    if (r < 0) return r;
 
     return sd_bus_reply_method_return(m, "o", obj_path);
 }
-- 
2.39.2

